swarm/storage/mru/lookup/lookup.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+/*
+Package lookup defines resource lookup algorithms and provides tools to place updates
+so they can be found
+*/
+package lookup
+
+const maxuint64 = ^uint64(0)
+const lowestLevel uint8 = 0
+const numLevels uint8 = 26
+const highestLevel = lowestLevel + numLevels - 1
+const defaultLevel = highestLevel
+
+// ReadFunc is a handler called by Lookup each time it attempts to find a value
+// It should return <nil> if a value is not found
+// It should return <nil> if a value is found, but its timestamp is higher than "now"
+// It should only return an error in case the handler wants to stop the
+// lookup process entirely.
+type ReadFunc func(epoch Epoch, now uint64) (interface{}, error)
+
+// NoClue is a hint that can be provided when the Lookup caller does not have
+// a clue about where the last update may be
+var NoClue = Epoch{}
+
+func getBaseTime(t uint64, level uint8) uint64 {
+	return t & (maxuint64 << level)
+}
+
+// Hint creates a hint based only on the last known update time
+func Hint(last uint64) Epoch {
+	return Epoch{
+		Time:  last,
+		Level: defaultLevel,
+	}
+}
+
+func getNextLevel(last Epoch, now uint64) uint8 {
+	mix := (last.Base() ^ now) | (1 << (last.Level - 1))
+	if mix > (maxuint64 >> (64 - highestLevel - 1)) {
+		return highestLevel
+	}
+	mask := uint64(1 << (highestLevel))
+
+	for i := uint8(highestLevel); i > lowestLevel; i-- {
+		if mix&mask != 0 {
+			return i
+		}
+		mask = mask >> 1
+	}
+	return 0
+}
+
+// GetNextEpoch returns the epoch where the next update should be located
+// according to where the previous update was
+// and what time it is now.
+func GetNextEpoch(last Epoch, now uint64) Epoch {
+	if last == NoClue {
+		return GetFirstEpoch(now)
+	}
+	level := getNextLevel(last, now)
+	return Epoch{
+		Level: level,
+		Time:  now,
+	}
+}
+
+// GetFirstEpoch returns the epoch where the first update should be located
+// based on what time it is now.
+func GetFirstEpoch(now uint64) Epoch {
+	return Epoch{Level: highestLevel, Time: now}
+}
+
+// Lookup finds the update with the highest timestamp that is smaller or equal than 'now'
+// It takes a hint which should be the epoch where the last known update was
```

Comment:
rather what the timestamp says. This gives a lower bound on which cells to look up

Reply:
The hint is not a lower bound. If the algorithm finds out the hint does not contain an update, it will then discard it. This is so you can provide approximate hints.
You say "cell", but the "cells" are called "epoch" in this algorithm.